src/arithmetic/algebraic_expression.c

```diff
@@ -305,95 +201,357 @@ void AlgebraicExpression_PrependTerm(AlgebraicExpression *ae, GrB_Matrix m, bool
-                if(!s) continue;
-                GrB_Matrix l = Graph_GetRelationMatrix(g, s->id);
-                GrB_Info info = GrB_eWiseAdd_Matrix_Semiring(m, NULL, NULL, Rg_structured_bool, m, l, NULL);                
+    op.operand = mat;
+    op.diagonal = false;
+    op.free = freeMatrix;
+    op.transpose = transpose;
+    return op;
+}
+
+AlgebraicExpression *AlgebraicExpression_Empty(void) {
+    return _AE_MUL(1);
+}
+
+/* In case edge a and b share a node:
+ * (a)-[E0]->(b)<-[e1]-(c)
+ * than the shared entity is returned
+ * if edges are disjoint, NULL is returned. */
+static Node* _SharedNode(const Edge *a, const Edge *b) {
+    assert(a && b);
+    if(a->src == b->src) return a->src;
+    if(a->src == b->dest) return a->src;
+    if(a->dest == b->src) return a->dest;
+    if(a->dest == b->dest) return a->dest;
+    return NULL;
+}
+
+static AlgebraicExpression* _AlgebraicExpression_From_Path(Edge **path, uint path_len, const AST *ast) {
+    assert(path && path_len > 0 && ast);
+
+    Edge *e = NULL;
+    AlgebraicExpressionOperand op;
+
+    // Construct expression.
+    AlgebraicExpression *exp = _AE_MUL(path_len * 2 - 1);  // (3) Node Edge Node.
+    // Count number of transposed edges.
+    int transposeCount = 0;
+
+    for(int i = 0; i < path_len; i++) {
+        e = path[i];
+        /* Determin if edge need to be transposed:
+         * Treating path as a chain
+         * we're aligning all edges to "point right"
+         * (A)-[E0]->(B)-[E0]->(C)-[E0]->(D).
+         * e.g. 
+         * (A)-[E0]->(B)<-[E1]-(C)-[E2]->(D)
+         * E1 will be transposed:
+         * (A)-[E0]->(B)-[E1]->(C)-[E2]->(D) */
+        bool transpose = false;
+        if(path_len > 1) {
+            if(i < path_len-1) {
+                Edge* follow = path[i+1];
+                Node *shared = _SharedNode(e, follow);
+                transpose = (e->dest != shared);
+            } else {
+                // Last edge.                
+                Edge* prev = path[i-1];
+                Node *shared = _SharedNode(prev, e);
+                transpose = (e->src != shared);
+            }
+            if(transpose) {
+                transposeCount++;
+                Edge_Reverse(e);
             }
-            mat = m;
         }
 
-        unsigned int hops = 1;
-        /* Expand fixed variable length edge */
-        if(astEdge->length && AST_LinkEntity_FixedLengthEdge(astEdge)) {
-            hops = astEdge->length->minHops;
+        // If src node has a label, multiple by label matrix.
+        if(e->src->label) {
+            op = _AlgebraicExpression_OperandFromNode(e->src);
+            AlgebraicExpression_AppendOperand(exp, op);
         }
 
+        // Add Edge matrix.
+        op = _AlgebraicExpression_OperandFromEdge(e, transpose, ast);
+
+        /* Expand fixed variable length edge */
+        unsigned int hops = (!Edge_VariableLength(e)) ? e->minHops : 1;
         for(int i = 0; i < hops; i++) {
```

Comment:
Use `j`; this overwrites outer loop variable `i` at 415.

Reply:
👍